torch/lib/THS/generic/THSTensorMath.c

```diff
@@ -5,6 +5,222 @@
+    THTensor_(free)(t_values_);
+  }
+}
+
+void THSTensor_(cadd)(THSTensor *r_, THSTensor *t, real value, THSTensor *src) {
+  if(!THSTensor_(isSameSizeAs)(t, src)) {
+    THError("cadd operands have incompatible sizes or dimension types");
+  }
+  THSTensor_(contiguous)(t);
+  THSTensor_(contiguous)(src);
+
+  if (src->nnz == 0) {
+    THSTensor_(copy)(r_, t);
+    return;
+  }
+  if (t->nnz == 0) {
+    THSTensor_(mul)(r_, src, value);
+    return;
+  }
+
+  // saving those because they can be overwritten when doing in-place operations
+  ptrdiff_t t_nnz = t->nnz, s_nnz = src->nnz, max_nnz = t_nnz + s_nnz;
+  long nDimI = THSTensor_(nDimensionI)(src);
+  long nDimV = THSTensor_(nDimensionV)(src);
+  THLongTensor *t_indices_ = THSTensor_(indices)(t);
+  THTensor *t_values_ = THSTensor_(values)(t);
+  THLongTensor *src_indices_ = THSTensor_(indices)(src);
+  THTensor *s_values_ = THSTensor_(values)(src);
+  THLongTensor *r_indices_ = THLongTensor_newWithSize2d(nDimI, max_nnz);
+  THTensor *r_values_ = THSTensor_(newValuesWithSizeOf)(s_values_, max_nnz);
+  THTensor_(zero)(r_values_);
+  // TODO handle case where src values is empty
+  THSTensor_(resizeAs)(r_, src);
+  THSTensor_(move)(r_, r_indices_, r_values_);
+
+  THTensor *srcBuffer = THTensor_(new)();
+  THTensor *dstBuffer = THTensor_(new)();
+  long cmp, d;
+  long r_i = 0, t_i = 0, s_i = 0;
+  while (t_i < t_nnz || s_i < s_nnz) {
+    if (t_i >= t_nnz) {
+      cmp = -1;
+    } else if (s_i >= s_nnz) {
+      cmp = 1;
+    } else {
+      cmp = 0;
+      for (d = 0; d < nDimI; d++) {
+        if (THTensor_fastGet2d(t_indices_, d, t_i) < THTensor_fastGet2d(src_indices_, d, s_i)) {
+          cmp = 1;
+          break;
+        }
+        if (THTensor_fastGet2d(t_indices_, d, t_i) > THTensor_fastGet2d(src_indices_, d, s_i)) {
+          cmp = -1;
+          break;
+        }
+      }
+    }
+    if (cmp >= 0) {
+      for (d = 0; d < nDimI; d++) {
+        THTensor_fastSet2d(r_indices_, d, r_i, THTensor_fastGet2d(t_indices_, d, t_i));
+      }
+      THSTensor_(addSlice)(dstBuffer, dstBuffer, srcBuffer, r_values_, r_values_, 1, t_values_, 0, r_i, r_i, t_i);
+      t_i++;
+    }
+    if (cmp <= 0) {
+      for (d = 0; d < nDimI; d++) {
+        THTensor_fastSet2d(r_indices_, d, r_i, THTensor_fastGet2d(src_indices_, d, s_i));
+      }
+      THSTensor_(addSlice)(dstBuffer, dstBuffer, srcBuffer, r_values_, r_values_, value, s_values_, 0, r_i, r_i, s_i);
+      s_i++;
+    }
+    r_i++;
+  }
+
+  r_->nnz = r_i;
+  r_->contiguous = 1;
+
+  THLongTensor_free(t_indices_);
+  THTensor_(free)(t_values_);
+  THLongTensor_free(src_indices_);
+  THTensor_(free)(s_values_);
+  THTensor_(free)(srcBuffer);
+  THTensor_(free)(dstBuffer);
+}
+
+void THSTensor_(csub)(THSTensor *r_, THSTensor *t, real value, THSTensor *src) {
+  THSTensor_(cadd)(r_, t, -value, src);
+}
+
+void THSTensor_(cmul)(THSTensor *r_, THSTensor *t, THSTensor *src) {
+  if(!THSTensor_(isSameSizeAs)(t, src)) {
```

Comment:
I think in `TH` and `THC` `c*` functions only check if the number of elements matches. But I'm fine with leaving it like that too.